### ISSUE TITLE
System Error on Python 3.6.a+ (bisected to 3 commits of upstream)

### ISSUE THREAD

**Issue by Carreau** [pytest-dev/pytest#1925]
Hi There, 

This is probably an upstream issue, and I would be happy to file a bug there, but I might need a bit of help to get a reproducible test case that does not involve py-test (which is awesome BTW)

I basically get the following system error while running the test-suite for https://github.com/xonsh/xonsh

```
Traceback (most recent call last):
  File "/Users/bussonniermatthias/dev/git-cpython/Lib/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Users/bussonniermatthias/dev/git-cpython/Lib/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.6/site-packages/pytest.py", line 17, in <module>
    raise SystemExit(pytest.main())
  File "/usr/local/lib/python3.6/site-packages/_pytest/config.py", line 57, in main
    return config.hook.pytest_cmdline_main(config=config)
  File "/usr/local/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/usr/local/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/usr/local/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/usr/local/lib/python3.6/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
    res = hook_impl.function(*args)
  File "/usr/local/lib/python3.6/site-packages/_pytest/main.py", line 125, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/usr/local/lib/python3.6/site-packages/_pytest/main.py", line 121, in wrap_session
    config._ensure_unconfigure()
  File "/usr/local/lib/python3.6/site-packages/_pytest/config.py", line 869, in _ensure_unconfigure
    fin()
  File "/usr/local/lib/python3.6/site-packages/_pytest/capture.py", line 79, in reset_capturings
    cap = self.__dict__.pop("_capturing", None)
SystemError: <built-in method pop of dict object at 0x10da54748> returned NULL without setting an error
```

Which definitively looks like it comes from CPython (`Objects/abstract.c` according to grep), which I have bisected down to 3 commits : 

```
commit e7ce8369093f2f685328f1a4dbcaf109bea97c03
Author: Benjamin Peterson <benjamin@python.org>
Date:   Thu Sep 8 09:50:08 2016 -0700

    rearrange to make gcc happy

commit 56b32ebb2aa3b760a0bd5f2c5951573a1dcb0f57
Author: Victor Stinner <victor.stinner@gmail.com>
Date:   Thu Sep 8 09:33:56 2016 -0700

    Add Py_MEMBER_SIZE macro

    Issue #27350: use Py_MEMBER_SIZE() macro to get the size of
    PyDictKeyEntry.dk_indices, rather than hardcoding 8.

commit 2330b19b3f6f0fd4e524f445eba05b8c1396736b
Author: Victor Stinner <victor.stinner@gmail.com>
Date:   Wed Sep 7 17:40:12 2016 -0700

    Implement compact dict

    Issue #27350: `dict` implementation is changed like PyPy. It is more compact
    and preserves insertion order.

    _PyDict_Dummy() function has been removed.

    Disable test_gdb: python-gdb.py is not updated yet to the new structure of
    compact dictionaries (issue #28023).

    Patch written by INADA Naoki.
```

(Note I'm using a git-cpython mirror as I'm not confortable with hg)

I've tried to reproduce it with a smaller example like:

```
class Foo:
    def set(self):
        self._capturing = 1
    def unset(self):
        cap = self.__dict__.pop("_capturing", None)
        return cap
f = Foo()
f.set()
f.unset()
f.unset()
```

So far without success... neither by creating `CaptureManager()` by hands. Am I missing anything that would mess with `self.__dict__`  in `_pytest/capture.py`? Any hints on debugging this further ?

Thanks.

**Comment by Carreau** [comment]
I've narrowed it down to a single parametric test in [this branch](https://github.com/Carreau/xonsh/tree/test-compact), one can see the [travis failure](https://travis-ci.org/Carreau/xonsh/builds/158659193) I'll narrow more later.

**Comment by RonnyPfannschmidt** [comment]
its could be related to their implementation of ordered dicts in dicts themselfes - i suggest checkering out if its still happening when opting out of that

**Comment by Carreau** [comment]
> its could be related to their implementation of ordered dicts in dicts themselfes - i suggest checkering out if its still happening when opting out of that

Yeah, I think it's related as the compact dict commit is one of the 3 I was able to bisect to. I've opened an issue of [bpo](http://bugs.python.org/issue28040) as well, but I'd like to figure out on a smaller test. It's also hard to "opt-out" except by rolling back the nightly.

Right now I've reduced it to

```
def test_show_cmd_numerate(capsys):
    """Verify that CLI history commands work."""
    pass
```

But almost any other change to any other part of the codebase make the bug go away... :-(

**Comment by RonnyPfannschmidt** [comment]
i suspect you hit a fairly uncommon case, im also not well versed in C, thus i cant help

**Comment by Carreau** [comment]
They seem to have figured that out upstream and have a proposed fix, I'll close this one the upstream issue is resolved if it does indeed fix it for me.

**Comment by haypo** [comment]
Oh wow, thank you for testing CPython development branch (default)! It helped us to detect bugs earlier, even before the beta scheduled for this weekend.

**Comment by Carreau** [comment]
> Oh wow, thank you for testing CPython development branch (default)! It helped us to detect bugs earlier, even before the beta scheduled for this weekend.

You should send thanks to travis-ci.org, they provide [nightly](https://docs.travis-ci.com/user/languages/python#Nightly-build-support) as one of the targets versions.
There is an increasing number of project on travis-ci supporting nightly.

I was even wondering if we couldn't use the ratio of fails on nightly vs fails on 3.5 across all travis builds to automatically detect regressions on CPython dev branch.

Thanks a lot for your work on CPython ! And looking Fwd to 3.6 !

**Comment by Carreau** [comment]
Things have been fixed upstream ! Thanks !

**Comment by haypo** [comment]
Yeah the issue should be fixed in Python 3.6 beta 1 released yesterday.
